File: ScriptBeeWebApp/src/Analysis/Application/Service.Analysis/FileBundler.cs
using System.Buffers.Binary;
using System.Text;
using ScriptBee.Common.CodeGeneration;

namespace ScriptBee.Service.Analysis;

public class FileBundler
{
    public async Task WriteToStream(
        IEnumerable<SampleCodeFile> files,
        Stream outputStream,
        CancellationToken cancellationToken = default
    )
    {
        foreach (var file in files)
        {
            var pathBytes = Encoding.UTF8.GetBytes(file.Name);
            var contentBytes = Encoding.UTF8.GetBytes(file.Content);

            var pathLengthBuffer = new byte[4];
            BinaryPrimitives.WriteInt32BigEndian(pathLengthBuffer, pathBytes.Length);
            await outputStream.WriteAsync(pathLengthBuffer, cancellationToken);

            await outputStream.WriteAsync(pathBytes, cancellationToken);

            var contentLengthBuffer = new byte[8];
            BinaryPrimitives.WriteInt64BigEndian(contentLengthBuffer, contentBytes.Length);
            await outputStream.WriteAsync(contentLengthBuffer, cancellationToken);

            await outputStream.WriteAsync(contentBytes, cancellationToken);
        }

        var endBuffer = new byte[4];
        BinaryPrimitives.WriteInt32BigEndian(endBuffer, 0);
        await outputStream.WriteAsync(endBuffer, cancellationToken);
    }
}

File: Integrations/sdks/python/scriptbee_analysis_sdk/file_bundler.py
import struct
from io import RawIOBase

from scriptbee_analysis_sdk.domain.code_generation import SampleCodeFile


class FileBundler:
    def write_to_stream(self, files: list[SampleCodeFile], stream: RawIOBase) -> None:
        for file in files:
            path_bytes = file.name.encode("utf-8")
            content_bytes = file.content.encode("utf-8")
            stream.write(struct.pack(">I", len(path_bytes)))
            stream.write(path_bytes)
            stream.write(struct.pack(">Q", len(content_bytes)))
            stream.write(content_bytes)
        stream.write(struct.pack(">I", 0))
